src/bytes/bytes_lemmas.gobra

```diff
@@ -0,0 +1,359 @@
+	assume forall j int :: { j in indices }{ s[j:j+len(sep)] } !(j in indices) && InRangeInc(j, 0, len(s) - len(sep)) && !SetContainsInRange(indices, j-len(sep), j) ==> s[j:j+len(sep)] != sep
+}
+
+
+
+// Repeat
+
+ghost
+ensures bytes.Repeat(b, 1) == b
+decreases
+func lemmaRepeatOnce(b seq[byte]) {
+	assert bytes.Repeat(b, 1) == b ++ bytes.Repeat(b, 0)
+}
+
+ghost
+requires n > 0
+ensures bytes.Repeat(b, n) == bytes.Repeat(b, n-1) ++ b
+decreases n
+func lemmaRepeatFlip(b seq[byte], n int) {
+	if n == 1 {
+		assert bytes.Repeat(b, n-1) == seq[byte]{}
+	} else {
+		assert bytes.Repeat(b, n) == b ++ bytes.Repeat(b, n-1)
+		lemmaRepeatFlip(b, n-1)
+		assert bytes.Repeat(b, n-1) == bytes.Repeat(b, n-2) ++ b
+	}
+}
+
+ghost
+requires n >= 0
+ensures  b ++ bytes.Repeat(b, n) == bytes.Repeat(b, n) ++ b
+decreases
+func lemmaRepeatCommutative(b seq[byte], n int) {
+	assert b ++ bytes.Repeat(b, n) == bytes.Repeat(b, n+1)
+	lemmaRepeatFlip(b, n+1)
+}
+
+ghost
+requires n >= 0
+ensures  bytes.Repeat(b, 2 * n) == bytes.Repeat(b, n) ++ bytes.Repeat(b, n)
+decreases n
+func lemmaRepeat_2n(b seq[byte], n int) {
+	if n != 0 {
+		assert bytes.Repeat(b, n) == b ++ bytes.Repeat(b, n - 1)
+		assert bytes.Repeat(b, n) ++ bytes.Repeat(b, n) == b ++ bytes.Repeat(b, n - 1) ++ b ++ bytes.Repeat(b, n - 1)
+		// order of repeating doesn't matter because it's always the same element that is repeated
+		assert b ++ bytes.Repeat(b, n - 1) ++ b ++ bytes.Repeat(b, n - 1) == b ++ (bytes.Repeat(b, n - 1) ++ b) ++ bytes.Repeat(b, n - 1)
+		lemmaRepeatCommutative(b, n - 1)
+		assert b ++ (bytes.Repeat(b, n - 1) ++ b) ++ bytes.Repeat(b, n - 1) == b ++ (b ++ bytes.Repeat(b, n - 1)) ++ bytes.Repeat(b, n - 1)
+		assert b ++ bytes.Repeat(b, n - 1) ++ b ++ bytes.Repeat(b, n - 1) == b ++ b ++ bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1)
+		lemmaRepeat_2n(b, n - 1)
+		assert bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1) == bytes.Repeat(b, 2 * (n - 1))
+		assert bytes.Repeat(b, 2 * n - 1) == b ++ bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1)
+		assert bytes.Repeat(b, 2 * n) == b ++ b ++ bytes.Repeat(b, n - 1) ++ bytes.Repeat(b, n - 1)
+	} else {
+		assert bytes.Repeat(b, 2 * 0) == bytes.Repeat(b, 0) ++ bytes.Repeat(b, 0)
+	}
+}
+
+ghost
+requires bp < lnb
+requires bp == lb * i
+requires lnb == lb * count
+requires 0 <= lb
+ensures  i < count
+decreases
+func lemmaStillHaveToRepeat(i, count, bp, lb, lnb int) {
+	if lb != 0 {
+		assert lb * i < lnb
+		assert lb * i < lb * count
+		LemmaMulInjLtLeftInv(i, count, lb)
+		assert i < count
+	}
+}
+
+
+
+
+
+ghost
+requires a == b
+requires count >= 0
+requires a >= 0
+requires nb[:MinInt(count, a) * len(s)] == bytes.Repeat(s, MinInt(count, a))
+ensures  nb[:MinInt(count, b) * len(s)] == bytes.Repeat(s, MinInt(count, b))
+decreases
+func lemmaSubstV(nb, s seq[byte], count, a, b int ) { }
```

Comment:
when is such a lemma useful? this is something that an smt solver should be able to do at any moment. I would expect that dropping this lemma would not cause any breakage

Reply:
Strangely, removing it and running verification with the following command fails: Might this be a bug?
```bash
gobra \
  "--backend" "SILICON" "--chop" "1" "--cacheFile" "/tmp/gobracache" "--assumeInjectivityOnInhale" \
  "--checkConsistency" "--mceMode=od" "--requireTriggers" "--moreJoins" "off" "-g" "/tmp/" \
  "-i" "/home/hyde/eth/6/bachelor/verified_go_stdlib/src/bytes/bytes.go" \
  "-i" "/home/hyde/eth/6/bachelor/verified_go_stdlib/src/bytes/bytes_lemmas.gobra" \
  "-I" "/home/hyde/eth/6/bachelor/verified_go_stdlib/src" \
  "--onlyFilesWithHeader" "--packageTimeout" "60s" "--disableNL"
```